### ISSUE TITLE
The `transpose` function is problematic

### ISSUE THREAD

**Issue by rgommers** [data-apis/array-api#228]
There are multiple problems, both related to adoption due to conflicting definitions, and because NumPy's `transpose` does something no one really wants (namely, swap _all_ axes).

The current array API `transpose` may not be correct:
- https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#transpose-x-axes-none says "swap all axes"
- https://github.com/data-apis/array-api/issues/147 claims the intent is to only swap the last two axes, to be consistent with all other `linalg` functions.

PyTorch names the `numpy.transpose` equivalent `permute`, and PyTorch's `transpose` is `np.swapaxes`. Relevant issues PyTorch issues:
  - https://github.com/pytorch/pytorch/issues/50275
  - https://github.com/pytorch/pytorch/issues/51280

Relevant NumPy issues:
- https://github.com/numpy/numpy/issues/9530
- https://github.com/numpy/numpy/issues/7495
- https://github.com/numpy/numpy/issues/13797 (note that there's `np.linalg.linalg.transpose` which is private but _does_ have the desired behavior)

There's multiple ways of dealing with this problem:
1. Accept the status quo
2. Remove `transpose` completely, and use `permute` or some other function name with the behavior we want by default (do the right thing for batches of 2-D matrices)
3. Add both, by adding a `matrix_transpose` function and a corresponding `.mT` (or `.MT`) attribute

I searched the JAX and TensorFlow issue trackers as well and, somewhat surprisingly, didn't find an issue about this topic.

**Comment by shoyer** [comment]
I think separating the functionality of `np.transpose()` into two separate functions/methods would make sense. As you note, there are two distinct use cases:

1. Swap the last two dimensions
2. Permute axes

The only real relation is that (1) can be implemented via (2).

From my perspective, the functionality (1) would be the best fit for the name "transpose", and (2) could use some other name, like "permute axes".

**Comment by kgryte** [comment]
During the consortium meeting on 2021-07-14, the following was considered a path forward:

1.  Limit the `.T` attribute on array instances to 2D. If an array instance has more than two dimensions, an error should be raised. This allows retaining the current behavior for `.T` for the primary use case of taking the transpose of a matrix, while removing the less desired behavior of reversing all axes.
2.  Introduce a `.mT` attribute on array instances which performs a batched matrix transpose for the last two dimensions.
3.  Remove the `transpose` function from the specification and replace with a `permute` API.
4.  Introduce a `matrix_transpose` function for performing a batched matrix transpose.

**Comment by shoyer** [comment]
> * Remove the `transpose` function from the specification and replace with a `permute` API. **PR**: [gh-247](https://github.com/data-apis/array-api/pull/247).

any thoughts on the name `permute` vs `permute_axes`?

I like `permute_axes` a little bit better, because I can imagine "permute" alternatively referring to array elements.

**Comment by rgommers** [comment]
Good point @shoyer, there's also https://numpy.org/devdocs/reference/random/generated/numpy.random.Generator.permuted.html which is quite similar. I don't have a strong preference, but `permute_axes` is indeed slightly clearer.

**Comment by kgryte** [comment]
Some other reference points:

-   MATLAB uses [`permute`](https://www.mathworks.com/help/matlab/ref/permute.html).
-   Julia uses [`permutedims`](https://docs.julialang.org/en/v1/base/arrays/#Base.permutedims)
-   PyTorch uses [`permute`](https://pytorch.org/docs/stable/generated/torch.Tensor.permute.html)

I am fine with either `permute` or `permute_axes`. Although, maybe we should be consistent with names. For example,

-   [`expand_dims`](https://github.com/data-apis/array-api/blob/a7a91fea22007c0c211ccbaa09a10c710e694620/spec/API_specification/manipulation_functions.md#expand_dimsx---axis)
-   while not currently included in the specification, [`swapaxes`](https://numpy.org/doc/stable/reference/generated/numpy.swapaxes.html).

In which case, to throw another hat in the ring, maybe `permute_dims`, if we want to be more explicit about what is being permuted?

**Comment by rgommers** [comment]
`permute_dims` seems quite reasonable too. If we consistently name keywords `axis`/`axes`, using `_dims` for function names and `ndim` for attribute is at least some kind of consistency.

**Comment by kgryte** [comment]
As [gh-245](https://github.com/data-apis/array-api/pull/245), [gh-246](https://github.com/data-apis/array-api/pull/246), [gh-247](https://github.com/data-apis/array-api/pull/247), and [gh-248](https://github.com/data-apis/array-api/pull/248) have all been merged, closing this out.
